### PR TITLE
Add <earlier_x.unit> relative date operator

### PR DESCRIPTION
**Pull request by sluc23**
Overview
----------------------------------------
Adds a new (configurable) relative date operator based on `earlier_x.[unit]` notation. Now this operator is not supported and only *1 unit*  fixed operators are available, as `earlier.day`, `earlier.week `, `earlier.month`, `earlier.quarter`, `earlier.year`.

`[unit]` is one of these options:
- day
- week
- month
- year
- quarter

With this PR a new operator can be created as `OptionValue` for `relative_date_filters` group like `earlier_6.month`, `earlier_3.year`, `earlier_5.week` to get everything before that relative date in the past

Before
----------------------------------------
Relative date operator `earlier_x.[unit]` is not processed

After
----------------------------------------
Custom date operators created in Civi as `OptionValues` with the format `earlier_x.[unit]` can be precessed.

Technical Details
----------------------------------------
UnitTest included

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
yay ='has-test'

@magnolia61 @MegaphoneJon either of you happy to take a look?

Ideally a docs update to this page ... https://docs.civicrm.org/user/en/latest/searching/relative-date-formats/

**Comment by sluc23** [comment]
agree with both of you @homotechsual @mattwire that rephrasing this operator name would be more clear.
It won't take much effort renaming it in my implementation, but what do we do with existing (in core) filters `earlier.year`, `earlier.quarter`, `earlier.month`, `earlier.week`, `earlier.day` ?

It doesn't seem feasible we can rename those without breaking existing reports, smart groups, searches, custom developments, etc, using them.. so would be a good idea to have 2 different name conventions for operators that have the same implementation/objective?

**Comment by sluc23** [comment]
just as a reference the method  `CRM_Utils_Date::relativeToAbsolute($relativeTerm, $unit)` is processing any operator with the format `operator_x.units`, first it breaks up from the "." to separate `operator_x` from `unit`, and after splits the "_" to get `operator` from `x`.. so any new name following this convention will be easy to implement, without messing up too much with `relativeToAbsolute()` (that function is complex.. wouldn't like to mess it up .. :S )

mention this because names like "earlier_then_4.week", "before_than_5.days",  will have a huge impact on the development..

So we need some kind of consensus here, let's see what others think about it.. @eileenmcnaughton @demeritcowboy @totten

**Comment by mattwire** [comment]
> That's fair @sluc23 :-)
> 
> Do the tokens get the previous x periods of data? if so I think `previous_x.unit` makes sense.
> 
> `previous_5.quarters`

@homotechsual No, it's everything from the beginning of time eg. "Everything before 1 year ago"

**Comment by sluc23** [comment]
any final decision if this PR needs some adjustments or could be merged? would be good to decide soon to avoid to rebase it in case `CRM_Utils_Date` class changes..

**Comment by eileenmcnaughton** [comment]
@sluc23 as I mentioned ['earlier' is an existing word](https://docs.civicrm.org/user/en/latest/searching/relative-date-formats/) - not a new word - so I think it makes sense to expand the existing terminlogy. In user facing places we should use more textual description that clarifies but I don't think this change warrants a complete re-visit of the code descriptors

image

The challenge is how does it fit with other code describers - I've copied them below, without the greater set - 

| value | Means |Defines|
| ---- | ---- |----|
| current.month | Current calendar month to-date |start & end|
| current.year | Current calendar year to-date |start & end|
| current.quarter | Current quarter to-date |start & end|
| current.week | Current week to-date |start & end|
| previous\_before.day | Day prior to yesterday |start & end|
| ending.year | Last 12 months including today |start & end|
| ending\_30.day | Last 30 days including today |start & end|
| ending\_60.day | Last 60 days including today |start & end|
| ending.week | Last 7 days including today |start & end|
| ending\_90.day | Last 90 days including today |start & end|
| previous\_before.month | Month prior to previous calendar month |start & end|
| previous\_2.month | Previous 2 calendar months |start & end|
| previous\_2.year | Previous 2 calendar years |start & end|
| previous\_2.day | Previous 2 days |start & end|
| previous\_2.quarter | Previous 2 quarters |start & end|
| previous\_2.week | Previous 2 weeks |start & end|
| previous.month | Previous calendar month |start & end|
| previous.year | Previous calendar year |start & end|
| previous.fiscal\_year | Previous fiscal year |start & end|
| previous.quarter | Previous quarter |start & end|
| previous.week | Previous week |start & end|
| previous\_before.quarter | Quarter prior to previous quarter |start & end|
| this.month | This calendar month |start & end|
| this.year | This calendar year |start & end|
| this.fiscal\_year | This fiscal year |start & end|
| this.quarter | This quarter |start & end|
| this.week | This week |start & end|
| less.month | To end of current calendar month |end
| less.year | To end of current calendar year |end
| less.quarter | To end of current quarter |end
| less.week | To end of current week |end
| previous\_before.week | Week prior to previous week |state and end|
| previous\_before.year | Year prior to previous calendar year |state and end|
| previous.day | Yesterday |state and end|
| earlier.month | To end of previous calendar month |end|
| earlier.year | To end of previous calendar year |end|
| earlier.quarter | To end of previous quarter |end|
| earlier.week | To end of previous week |end|
| earlier.day | To end of yesterday |end|
|**earlier_3.day| To end of 3 days ago.|end|
|earlier_3.week| To end of 3 weeks ago.|end|
|earlier_3.month| To end of 3 months ago.|end|
|earlier_3.quarter| To end of last day of 3 quarters ago.|end|
|earlier_3.year| To end of 3 years ago.|end|**

Looking at this we have just 2  describers that specify only the end date

- earlier
- less

`less` covers an end point in the future and `earlier` a point in the past - so in that sense `earlier` IS the best exisitng very. The test for me as to whether it works is whether the existing values can be rewritren with '1' - ie

is 

| value | Means |Defines|
| ---- | ---- |----|
| earlier.year | To end of previous calendar year |end|
| earlier.quarter | To end of previous quarter |end|

compatible with 

| value | Means |Defines|
| ---- | ---- |----|
| earlier_1.year | To end of 1 years ago |end|
|earlier_1.quarter| To end of last day of 1 quarters ago.|end|

I think it is - although the language of course reads oddly once you put a 1 in it

So I'm inclined to think that this is the right approach - but I haven't checked the code / r-run at this stage

**Comment by eileenmcnaughton** [comment]
let's re-run tests - test this please

**Comment by eileenmcnaughton** [comment]
Hmm - I am getting that test bork on another PR - @seamuslee001 any idea?

**Comment by seamuslee001** [comment]
Jenkins re test this please

**Comment by seamuslee001** [comment]
@eileenmcnaughton see infrastructure channel in mattermost

File: CRM/Utils/Date.php [hunk 1]
```diff
@@ -1258,6 +1258,16 @@ public static function relativeToAbsolute($relativeTerm, $unit) {
                   $from['Y'] -= ($relativeTermSuffix - 1);
                 }
                 break;
+
+              case 'earlier':
+                $to['d'] = $now['mday'];
+                $to['M'] = $now['mon'];
```
Comment (demeritcowboy) [inline CRM/Utils/Date.php:1264]: I don't think this is correct. Compare to line 1157. If I do earlier_2.year it says Dec 9, 2019, but it should be Dec 31, 2019.

**Comment by sluc23** [comment]
honestly I've got very disappointed about this PR.. I've spent few hours adding a new functionality, **consistent** with existing "earlier" terminology, and it got stuck in large discussion about proper "wording".. anyway we added this as a patch in our installations.. thanks everybody for the feedback!

**Comment by homotechsual** [comment]
FWIW I think it's worth bringing this back up-to current master if you're willing and getting it merged - it's an area where this might not be the perfect implementation but it's better than what we have.

I didn't mean to reopen this - apologies.

**Comment by eileenmcnaughton** [comment]
@sluc23 @homotechsual  - yep the blocking issue in the end is the technical one per https://github.com/civicrm/civicrm-core/pull/21858#pullrequestreview-828100454 (plus it needs a rebase - if you want to rebase @sluc23 I can re-open)

**Pull request closed without merging**